main/Repl.hs

```diff
@@ -94,83 +181,138 @@ exec
+    -- and try again.
+    --
+    -- This allows us to handle assignments like @a = 42@
+    -- which get turned into @{ a = 42; }@
+    parseExprOrBinding i =
+      case parseNixTextLoc i of
+        Success expr -> (Success expr, False)
+        Failure e    ->
+          case parseNixTextLoc $ toAttrSet i of
+            Failure _  -> (Failure e, False) -- return the first parsing failure
+            Success e' -> (Success e', True)
+
+    toAttrSet i = "{" <> i <> (if Data.Text.isSuffixOf ";" i then mempty else ";") <> "}"
 
 cmd
   :: (MonadNix e t f m, MonadIO m)
   => String
   -> Repl e t f m ()
 cmd source = do
-  val <- exec True (Text.pack source)
+  mVal <- exec True (Data.Text.pack source)
+  case mVal of
+    Nothing -> return ()
+    Just val -> printValue val
+
+printValue :: (MonadNix e t f m, MonadIO m)
+           => NValue t f m
+           -> Repl e t f m ()
+printValue val = do
+  cfg <- replCfg <$> get
   lift $ lift $ do
-    opts :: Nix.Options <- asks (view hasLens)
     if
-      | strict opts -> liftIO . print . prettyNValue =<< normalForm val
-      | values opts -> liftIO . print . prettyNValueProv =<< removeEffects val
-      | otherwise   -> liftIO . print . prettyNValue =<< removeEffects val
+      | cfgStrict cfg -> liftIO . print . prettyNValue =<< normalForm val
+      | cfgValues cfg -> liftIO . print . prettyNValueProv =<< removeEffects val
+      | otherwise     -> liftIO . print . prettyNValue =<< removeEffects val
+
 -------------------------------------------------------------------------------
 -- Commands
 -------------------------------------------------------------------------------
 
 -- :browse command
-browse :: MonadNix e t f m => [String] -> Repl e t f m ()
+browse :: (MonadNix e t f m, MonadIO m)
+       => String
+       -> Repl e t f m ()
 browse _ = do
   st <- get
-  undefined
-  -- liftIO $ mapM_ putStrLn $ ppenv (tyctx st)
+  forM_ (Data.HashMap.Lazy.toList $ replCtx st) $ \(k, v) -> do
+    liftIO $ putStr $ Data.Text.unpack $ k <> " = "
+    printValue v
 
 -- :load command
 load
   :: (MonadNix e t f m, MonadIO m)
-  => [String]
+  => String
   -> Repl e t f m ()
 load args = do
-  contents <- liftIO $ Text.readFile (unwords args)
+  contents <- liftIO
+    $ Data.Text.IO.readFile
+    $ Data.Text.unpack
+    $ Data.Text.strip
+    $ Data.Text.pack args
   void $ exec True contents
 
 -- :type command
 typeof
   :: (MonadNix e t f m, MonadIO m)
-  => [String]
+  => String
   -> Repl e t f m ()
 typeof args = do
-  st  <- get
-  val <- case M.lookup line (tmctx st) of
-    Just val -> return val
-    Nothing  -> exec False line
-  str <- lift $ lift $ showValueType val
-  liftIO $ putStrLn str
-  where line = Text.pack (unwords args)
+  st <- get
+  mVal <- case Data.HashMap.Lazy.lookup line (replCtx st) of
+    Just val -> return $ Just val
+    Nothing  -> do
+      exec False line
+
+  maybe (pure ()) (\x -> (lift . lift . showValueType $ x) >>= liftIO . putStrLn) mVal
```

Comment:
The last puzzler I'm going to point out. ;)

I think it could be written roughly like this:
```suggestion
  for_ mVal $ \val -> do
    s <- lift . lift . showValueType $ val
    liftIO $ putStrLn s
```

Reply:
Neat! Thank you!